### PR TITLE
ENT-3455: Improve logging of ACL errors

### PR DESCRIPTION
**Pull request by craigcomstock**
Changelog: Title

### CODE REVIEW

**Comment by craigcomstock** [comment]
@jimis @olehermanse maybe? for review?

File: cf-agent/acl_posix.c [hunk 1]
```diff
@@ -339,15 +341,30 @@ static int CheckPosixLinuxACEs(EvalContext *ctx, Rlist *aces, AclMethod method,
 
             if (!DONTDO)
             {
+                int last = -1;
+                acl_text_str = acl_to_any_text(acl_new, NULL, ',', 0);
+                Log(LOG_LEVEL_VERBOSE, "ACL: new acl is `%s'", acl_text_str);
```
Comment (jimis) [inline cf-agent/acl_posix.c]: Shouldn't this be LOG_LEVEL_DEBUG?

File: cf-agent/acl_posix.c [hunk 2]
```diff
@@ -339,15 +341,30 @@ static int CheckPosixLinuxACEs(EvalContext *ctx, Rlist *aces, AclMethod method,
...
+
+                if ((retv = acl_check(acl_new, &last)) != 0)
+                {
+                    Log(LOG_LEVEL_ERR, "ACL: acl_check(%s): %s at entry %d. Refer to acl_check man page for validity rules.",
+			    acl_text_str,
+			    acl_error(retv),
+			    last);
+		    acl_free(acl_text_str);
+		    return false;
```
Comment (jimis) [inline cf-agent/acl_posix.c]: Free the rest of `acl_*` variables?

**Comment by craigcomstock** [comment]
oops... foobarred the rebase... working on it :)

**Comment by craigcomstock** [comment]
`bundle agent main
{
    files:
        "./test.tmp"
        create => "true",
        acl => template;
}

body acl template
{
    acl_method => "overwrite";
    aces => {
        "user:*:rwx",
        "user:craig:rwx",
        "group:*:rwx",
        "group:craig:rwx",
        "all:rwx"
    };
}

~
`

`craig@other:~/cfengine/core/tests/acceptance$ ../../cf-agent/cf-agent -f ./test.cf
   error: ACL: acl_check(user:craig:rwx,group:craig:rwx,mask::rwx): Missing or wrong entry at entry 0. Refer to acl_check man page for validity rules.
   error: Failed checking access ACL on ./test.tmp
   error: Promise belongs to bundle 'main' in file './test.cf' near line 4
   error: ACL: acl_``

File: cf-agent/acl_posix.c [hunk 3]
```diff
@@ -339,15 +341,33 @@ static int CheckPosixLinuxACEs(EvalContext *ctx, Rlist *aces, AclMethod method,
...
+                if ((retv = acl_check(acl_new, &last)) != 0)
+                {
+                    Log(LOG_LEVEL_ERR, "ACL: acl_check(%s): %s at entry %d. Refer to acl_check man page for validity rules.",
+			            acl_text_str,
+        			    acl_error(retv),
+        			    last);
+                    acl_free((void *) acl_existing);
+                    acl_free((void *) acl_tmp);
+                    acl_free((void *) acl_new);
+         		    acl_free(acl_text_str);
```
Comment (jimis) [inline cf-agent/acl_posix.c]: `acl_free()` is not for strings. Just use `free()`.

File: cf-agent/acl_posix.c [hunk 4]
```diff
@@ -339,15 +341,33 @@ static int CheckPosixLinuxACEs(EvalContext *ctx, Rlist *aces, AclMethod method,
 
             if (!DONTDO)
             {
+                int last = -1;
+                acl_text_str = acl_to_any_text(acl_new, NULL, ',', 0);
+                Log(LOG_LEVEL_DEBUG, "ACL: new acl is `%s'", acl_text_str);
+
+                if ((retv = acl_check(acl_new, &last)) != 0)
+                {
+                    Log(LOG_LEVEL_ERR, "ACL: acl_check(%s): %s at entry %d. Refer to acl_check man page for validity rules.",
```
Comment (jimis) [inline cf-agent/acl_posix.c]: Try to improve indentation and wrap better, if possible at column 78 max. And make sure that messages higher than log level debug are human-readable. For example this one:

```
                    Log(LOG_LEVEL_ERR, "Invalid ACL in '%s' at index %d (acl_check: %s)"
                        acl_text_str, last, acl_error(retv)); 
```

File: cf-agent/acl_posix.c [hunk 5]
```diff
@@ -339,15 +341,33 @@ static int CheckPosixLinuxACEs(EvalContext *ctx, Rlist *aces, AclMethod method,
...
+			            acl_text_str,
+        			    acl_error(retv),
+        			    last);
+                    acl_free((void *) acl_existing);
```
Comment (jimis) [inline cf-agent/acl_posix.c]: No casting for `void *` parameters, in C `void *` casting happens implicitly. It's a C++ thing that is not needed in C, and even considered harmful by some. I know the old code does it, but if possible change it.

File: cf-agent/acl_posix.c (hunk 3 above)
Comment (craigcomstock) [reply cf-agent/acl_posix.c]: I took the hint from acl_to_text(3) ```This function allocates any memory necessary to contain the string and returns a pointer to the string.  The caller should free
     any releasable memory, when the new string is no longer required, by calling acl_free(3) with the (void*)char returned by
     acl_to_text() as an argument.
``` Same difference I suppose but it seems more proper to use acl_free()?

**Comment by jimis** [comment]
You are right, ignore this comment then!

File: cf-agent/acl_posix.c (hunk 5 above)
Comment (craigcomstock) [reply cf-agent/acl_posix.c]: Do you have a reference for reading so I can get some background? Something like this? https://gustedt.wordpress.com/2014/04/02/dont-use-casts-i/

File: cf-agent/acl_posix.c (hunk 3 above)
Comment (jimis) [reply cf-agent/acl_posix.c]: You are right, ignore this comment then! (copied misplaced comment below)

File: cf-agent/acl_posix.c (hunk 5 above)
Comment (jimis) [reply cf-agent/acl_posix.c]: First let me say that there are valid uses of casts. But when you use a cast you are basically telling the compiler "don't warn me of anything, I know what I'm doing". So in general it's good practice to keep the casts at a minimum and fix compiler warnings at their root cause, if possible (again, sometimes casts are needed).

The link you posted is quite good. There were also some more real problems with legacy C compilers: if you casted the return of a malloc() for example and had forgotten to include the proper header file, you might get a runtime error. Nice posts [here](https://stackoverflow.com/questions/1565496/specifically-whats-dangerous-about-casting-the-result-of-malloc).

**Comment by craigcomstock** [comment]
updated messages:

body acl template
{
    acl_method => "overwrite";
    aces => {
#        "user:*:rwx",
        "user:craig:rwx",
#        "group:*:rwx",
        "group:craig:rwx",
#        "all:rwx"
    };
}

gives the following because you can't have a user:craig:rwx without the user:*:rwx entry that is required.

2017-12-06T11:16:04-0600    error: Invalid ACL in 'user:craig:rwx,group:craig:rwx,mask::rwx' at index 0 (acl_check: Missing or wrong entry)
2017-12-06T11:16:04-0600    error: Failed checking access ACL on ./test.tmp
2017-12-06T11:16:04-0600    error: Promise belongs to bundle 'main' in file './craig-acl-test.cf' near line 4

Still the error doesn't quite fully describe the problem but I suppose it is better. :(

**Pull request merged**